src/librustc/ty/layout.rs

```diff
@@ -706,13 +707,35 @@ impl<'a, 'tcx> LayoutCx<'tcx, TyCtxt<'a, 'tcx, 'tcx>> {
                         } else {
                             align = align.max(field.align);
                         }
+
+                        if field.abi != Abi::Uninhabited {
+                            if abi == Abi::Uninhabited || size == Size::ZERO {
+                                abi = field.abi.clone();
+                            } else if field.size != Size::ZERO && abi != field.abi  {
+                                abi = Abi::Aggregate{ sized: true };
+                            }
+                        }
+
                         size = cmp::max(size, field.size);
                     }
 
+                    // Use scalar_unit to reset the valid range to the maximal one
+                    // since an union might not be initialised
```

Comment:
Need to do this before comparing `field.abi` above with `abi`, so we never compare ranges.